docs/error-type.md

```diff
@@ -0,0 +1,94 @@
+# Solidity `error` type
+
+## Motivation
+
+We want to support typed errors in a syntactically equivalent and convenient way like
+the programmer can define `event` types.
+
+Example:
+
+```solidity
+contract C {
+	event HighestBidIncreased(address bidder, uint amount);
+	error BiddingFailed(string reason);
+
+	function bid() public payable {
+		try attemptBid() {
+			emit HighestBidIncreased(msg.sender, msg.value); // Triggering event
+		} catch BiddingFailed(reason) {
+			// handle error with given reason
+		}
+    }
+
+	function attemptBid() {
+		// assert() should now support a custom secondary parameter to throw custom errors upon
+		failed expectations.
+		assert(middingPossible, BiddingFailed("Bidding impossible"));
+	}
+}
+```
+
+## Syntax
+
+```bnf
+errorDefinition
+  : 'error' identifier eventParameterList AnonymousKeyword? ';' ;
+```
+
+## Semantic
+
+Errors can be instanciated whereever a variable of a type can be declared.
```

Comment:
Do we really want variables of error type?

Reply:
I think it _may_ potentially be easier to not allow this in the beginning (?), but I had a usecase in my mind where the error type (as value) is being passed as function parameter to allow a little bit more flexibility.

I think we can just start with the basic case, and then (if people reqeust it), still extend the scope of error types.